File: Database Project/sources/java/DatabaseHelper.java
//Database Systems (Module IDS) 

import java.sql.*;
import java.util.ArrayList;


// The DatabaseHelper class encapsulates the communication with our database
class DatabaseHelper {
    // Database connection info
    private static final String DB_CONNECTION_URL = "jdbc:oracle:thin:@oracle19.cs.univie.ac.at:1521:orclcdb";
    private static final String USER = "xxx"; 
    private static final String PASS = "xxx";

    private static Statement stmt;
    private static Connection con;

    //CREATE CONNECTION
    DatabaseHelper() {
        try {
            //Loads the class into the memory
            //Class.forName(CLASSNAME);

            // establish connection to database
            con = DriverManager.getConnection(DB_CONNECTION_URL, USER, PASS);
            stmt = con.createStatement();

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    //INSERT INTO USERIN
    void insertIntoUserin(String username, String country, String fav_genre) {
        try {
            String sql = "INSERT INTO USERIN (USERNAME,LAND, FAV_GENRE) VALUES ('" + username + "','" +
                    country +
                    "', '" +
                    fav_genre +
                    "')";
            stmt.execute(sql);
        } catch (Exception e) {
            System.err.println("Error at: insertIntoUserIn\nmessage: " + e.getMessage());
        }
    }
    
    //INSERT INTO PLAYLIST
    void insertIntoPlaylist(String username, String listname, int songcount, String genre) {
        try {
            String sql = "INSERT INTO PLAYLIST (USERNAME,LISTNAME, SONG_ANZAHL,GENRE) VALUES ('" + username + "','" +
                    listname +
                    "', '" +
                    songcount +
                    "','" + genre + "')";
            stmt.execute(sql);
        } catch (Exception e) {
            System.err.println("Error at: insertIntoPlaylist\nmessage: " + e.getMessage());
        }
    }
    
    //INSERT INTO SONG
    void insertIntoSong( String songname, String length) {
        try {
            String sql = "INSERT INTO SONG (SONGNAME, LAENGE) VALUES ('" +
                    songname +
                    "', '" +
                    length + "')";
            stmt.execute(sql);
        } catch (Exception e) {
            System.err.println("Error at: insertIntoWebSong\nmessage: " + e.getMessage());
        }
    }
    
    //INSERT INTO SingleKuenstler
    void insertIntoSingleKuenstler( String k_name, int k_age) {
        try {
            String sql = "INSERT INTO SINGLEKUENSTLER (KNAME,KALTER) VALUES ('" +
                    k_name +
                    "', '" +
                    k_age + "')";
            stmt.execute(sql);
        } catch (Exception e) {
            System.err.println("Error at: insertIntoSingleKuenstler\nmessage: " + e.getMessage());
        }
    }
    
    //INSERT INTO BandKuenstler
    void insertIntoBandKuenstler(String k_name) {
        try {
            String sql = "INSERT INTO BANDKUENSTLER (KNAME) VALUES ('" + k_name + "')";
            stmt.execute(sql);
        } catch (Exception e) {
            System.err.println("Error at: insertIntoBandKuenstler\nmessage: " + e.getMessage());
        }
    }

    //INSERT INTO Zeitpunkt
    void insertIntoZeitpunkt(String date, String place) {
        try {
            String sql = "INSERT INTO ZEITPUNKT (DATUM, ORT) VALUES ('" + date + "','" + place +  "')";
            stmt.execute(sql);
        } catch (Exception e) {
            System.err.println("Error at: insertIntoZeitpunkt\nmessage: " + e.getMessage());
        }
    }
    
    //INSERT INTO SingleVeroeffentlicht
    void insertIntoSingleVeroeffentlicht(int kid, int songid) {
        try {
            String sql = "INSERT INTO SINGLEVEROEFFENTLICHT (KID,SONGID) VALUES ('" + kid + "','" +
                    songid +
                    "')";
            stmt.execute(sql);
        } catch (Exception e) {
            System.err.println("Error at: insertIntoSingleVeroeffentlicht\nmessage: " + e.getMessage());
        }
    }
    
    //INSERT INTO BandVeroeffentlicht
    void insertIntoBandVeroeffentlicht(int kid, int songid) {
        try {
            String sql = "INSERT INTO BANDVEROEFFENTLICHT (KID,SONGID) VALUES ('" + kid + "','" +
                    songid +
                    "')";
            stmt.execute(sql);
        } catch (Exception e) {
            System.err.println("Error at: insertIntoBandVeroeffentlicht\nmessage: " + e.getMessage());
        }
    }
    
    //INSERT INTO Folgt
    void insertIntoFolgt(String username1, String username2) {
        try {
            String sql = "INSERT INTO FOLGT (USERNAME,USERNAME2) VALUES ('" + username1 + "','" +
                    username2 +
                    "')";
            stmt.execute(sql);
        } catch (Exception e) {
            System.err.println("Error at: insertIntoFolgt\nmessage: " + e.getMessage());
        }
    }
    
    
    //INSERT INTO Hoert
    void insertIntoHoert(String username, String date, String place, int songid) {
        try {
            String sql = "INSERT INTO HOERT (USERNAME, DATUM, ORT, SONGID) VALUES ('" + username + "','" +
            		date + "','" + place + "','" + songid + 
                    "')";
            stmt.execute(sql);
        } catch (Exception e) {
            System.err.println("Error at: insertIntoHoert\nmessage: " + e.getMessage());
        }
    }
    
    
    
    //SELECT USERNAME FROM USER
    ArrayList<String> selectUsernameFromWebUser() {
        ArrayList<String> usernames = new ArrayList<>();

        try {
            ResultSet rs = stmt.executeQuery("SELECT USERNAME FROM USERIN");
            while (rs.next()) {
                usernames.add(rs.getString("USERNAME"));
            }
            rs.close();
        } catch (Exception e) {
            System.err.println(("Error at: selectPersonIsFromPerson\n message: " + e.getMessage()).trim());
        }
        return usernames;
    }
    
    //SELECT SONGID FROM SONG
    ArrayList<Integer> selectSongIDfromSong() {
        ArrayList<Integer> songIDs = new ArrayList<>();

        try {
            ResultSet rs = stmt.executeQuery("SELECT SONGID FROM SONG");
            while (rs.next()) {
                songIDs.add(rs.getInt("SONGID"));
            }
            rs.close();
        } catch (Exception e) {
            System.err.println(("Error at: selectPersonIsFromPerson\n message: " + e.getMessage()).trim());
        }
        return songIDs;
    }
    
    //SELECT ARTISTID FROM SINGLEARTIST
    ArrayList<Integer> selectArtistIDfromSingleArtist() {
        ArrayList<Integer> artistIDs = new ArrayList<>();

        try {
            ResultSet rs = stmt.executeQuery("SELECT KID FROM SINGLEKUENSTLER");
            while (rs.next()) {
                artistIDs.add(rs.getInt("KID"));
            }
            rs.close();
        } catch (Exception e) {
            System.err.println(("Error at: selectPersonIsFromPerson\n message: " + e.getMessage()).trim());
        }
        return artistIDs;
    }
    
    //SELECT ARTISTID FROM BANDARTIST
    ArrayList<Integer> selectArtistIDfromBandArtist() {
        ArrayList<Integer> artistIDs = new ArrayList<>();

        try {
            ResultSet rs = stmt.executeQuery("SELECT KID FROM BANDKUENSTLER");
            while (rs.next()) {
                artistIDs.add(rs.getInt("KID"));
            }
            rs.close();
        } catch (Exception e) {
            System.err.println(("Error at: selectPersonIsFromPerson\n message: " + e.getMessage()).trim());
        }
        return artistIDs;
    }

    //SELECT ZEITPUNKT
    ArrayList<Zeitpunkt> selectZeitpunkt() {
        ArrayList<Zeitpunkt> zeitpunkt = new ArrayList<>();

        try {
            ResultSet rs = stmt.executeQuery("SELECT DATUM,ORT FROM ZEITPUNKT");
            while (rs.next()) {
                String datum = rs.getString("DATUM");
                String ort = rs.getString("ORT");
                Zeitpunkt help = new Zeitpunkt(datum,ort);
                zeitpunkt.add(help);
            }
            rs.close();
        } catch (Exception e) {
            System.err.println(("Error at: selectPersonIsFromPerson\n message: " + e.getMessage()).trim());
        }
        return zeitpunkt;
    }
    



    public void close()  {
        try {
            stmt.close(); //clean up
            con.close();
        } catch (Exception ignored) {
        }
    }
}

File: Database Project/sources/php/DatabaseHelper.php
<?php

class DatabaseHelper
{
    // Since the connection details are constant, define them as const
    // We can refer to constants like e.g. DatabaseHelper::username
    const username = 'xxx'; 
    const password = 'xxx';
    const con_string = 'oracle19.cs.univie.ac.at:1521/orclcdb'; 

    protected $conn;

    // Create connection in the constructor
    public function __construct()
    {
        try {
            $this->conn = oci_connect(
                DatabaseHelper::username,
                DatabaseHelper::password,
                DatabaseHelper::con_string
            );

            //check if the connection object is != null
            if (!$this->conn) {
                // die(String(message)): stop PHP script and output message:
                die("DB error: Connection can't be established!");
            }

        } catch (Exception $e) {
            die("DB error: {$e->getMessage()}");
        }
    }

    // Used to clean up
    public function __destruct()
    {
        // clean up
        oci_close($this->conn);
    }

    
    public function selectUsers($username, $land, $fav_genre)
    {

        $sql = "SELECT * FROM USERIN
            WHERE USERNAME LIKE '%{$username}%'
              AND upper(LAND) LIKE upper('%{$land}%')
              AND upper(FAV_GENRE) LIKE upper('%{$fav_genre}%')
            ORDER BY USERNAME ASC";


        $statement = oci_parse($this->conn, $sql);

        oci_execute($statement);

        oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);

        //clean up;
        oci_free_statement($statement);

        return $res;
    }

    public function selectSongs($songid, $songname, $laenge)
    {
    
        if($songid=='') {
        $sql = "SELECT * FROM SONG
            WHERE SONGID LIKE '%{$songid}%'
              AND upper(SONGNAME) LIKE upper('%{$songname}%')
              AND upper(LAENGE) LIKE upper('%{$laenge}%')
            ORDER BY SONGID ASC";
        }
        if($songid!='') {
            $sql = "SELECT * FROM SONG
            WHERE SONGID ='{$songid}'
              AND upper(SONGNAME) LIKE upper('%{$songname}%')
              AND upper(LAENGE) LIKE upper('%{$laenge}%')
            ORDER BY SONGID ASC";
        }

       
        $statement = oci_parse($this->conn, $sql);

        
        oci_execute($statement);

        oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);

        //clean up;
        oci_free_statement($statement);

        return $res;
    }

    public function selectArtists()
    {

        $sql = "SELECT * FROM SINGLEKUENSTLER";

        
        $statement = oci_parse($this->conn, $sql);

        // Executes the statement
        oci_execute($statement);

       
        oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);

        //clean up;
        oci_free_statement($statement);

        return $res;
    }

    public function selectBands()
    {

        $sql = "SELECT * FROM BANDKUENSTLER";

        $statement = oci_parse($this->conn, $sql);

        // Executes the statement
        oci_execute($statement);

        oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);

        //clean up;
        oci_free_statement($statement);

        return $res;
    }
	

    public function insertIntoPerson($name, $surname)
    {
        $sql = "INSERT INTO PERSON (NAME, SURNAME) VALUES ('{$name}', '{$surname}')";

        $statement = oci_parse($this->conn, $sql);
		$success = oci_execute($statement) && oci_commit($this->conn);
        oci_free_statement($statement);
        return $success;
    }

    public function insertIntoUser($username, $land, $fav_genre)
    {
        $sql = "INSERT INTO USERIN (USERNAME, LAND, FAV_GENRE) VALUES ('{$username}', '{$land}','{$fav_genre}')";

        $statement = oci_parse($this->conn, $sql);
		$success = oci_execute($statement) && oci_commit($this->conn);
        oci_free_statement($statement);
        return $success;
    }

    public function insertIntoSong($songname, $laenge)
    {
        $sql = "INSERT INTO SONG (SONGNAME, LAENGE) VALUES ('{$songname}','{$laenge}')";

        $statement = oci_parse($this->conn, $sql);
		$success = oci_execute($statement) && oci_commit($this->conn);
        oci_free_statement($statement);
        return $success;
    }

   

    public function deleteUser($username)
    {
        if($username=='') {
            $success = false;
            return $success;
        }
        
        $sql = "DELETE FROM USERIN WHERE USERNAME = '{$username}'";

        $statement = oci_parse($this->conn, $sql);
		$success = oci_execute($statement) && oci_commit($this->conn);

        if (oci_num_rows($statement) == 0) {
            $success = false;
        }
        oci_free_statement($statement);


        return $success;
    }


    public function deleteSong($songid)
    {
        if($songid=='') {
            $success = false;
            return $success;
        }
        
        $sql = "DELETE FROM SONG WHERE SONGID = '{$songid}'";

        $statement = oci_parse($this->conn, $sql);
		$success = oci_execute($statement) && oci_commit($this->conn);

        if (oci_num_rows($statement) == 0) {
            $success = false;
        }
        oci_free_statement($statement);


        return $success;
    }

    public function updateUser($username, $land, $fav_genre)
    {
        $sql = "UPDATE USERIN SET land = '{$land}', fav_genre = '{$fav_genre}' WHERE USERNAME = '{$username}'";

        $statement = oci_parse($this->conn, $sql);
		$success = oci_execute($statement) && oci_commit($this->conn);

        if (oci_num_rows($statement) == 0) {
            $success = false;
        }
        oci_free_statement($statement);
        return $success;
    }

    public function updateSong($songid, $songname, $laenge)
    {
        $sql = "UPDATE SONG SET SONGNAME = '{$songname}', LAENGE = '{$laenge}' WHERE SONGID = '{$songid}'";

        $statement = oci_parse($this->conn, $sql);
		$success = oci_execute($statement) && oci_commit($this->conn);

        if (oci_num_rows($statement) == 0) {
            $success = false;
        }
        oci_free_statement($statement);
        return $success;
    }


    public function getPlaylistsOfUser($username)
    {
        $sql = "SELECT * FROM PLAYLIST
            WHERE USERNAME = '{$username}'";

        $statement = oci_parse($this->conn, $sql);

        oci_execute($statement);

        oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);

        //clean up;
        oci_free_statement($statement);

        return $res;
    }


    public function getSongsOfArtist($artistid)
    {
        $sql = "SELECT * FROM ARTISTSONGS
            WHERE KID = '{$artistid}'";

        $statement = oci_parse($this->conn, $sql);

        oci_execute($statement);

        oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);

        //clean up;
        oci_free_statement($statement);

        return $res;
    }

    public function getSongsOfBand($artistid)
    {
        $sql = "SELECT * FROM BANDSONGS
            WHERE KID = '{$artistid}'";

        $statement = oci_parse($this->conn, $sql);

        oci_execute($statement);

        oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);

        //clean up;
        oci_free_statement($statement);

        return $res;
    }

    public function insertFollower($username1, $username2)
    {
        $errorcode = 1;
        $friendnum = 0;
        $sql = 'BEGIN InsertFollower(:username1, :username2, :errorcode, :friendnum); END;';
        $statement = oci_parse($this->conn, $sql);
    
        oci_bind_by_name($statement, ':username1', $username1);
        oci_bind_by_name($statement, ':username2', $username2);
        oci_bind_by_name($statement, ':errorcode', $errorcode);
        oci_bind_by_name($statement, ':friendnum', $friendnum);

        oci_execute($statement);

        oci_free_statement($statement);



        return ['errorcode' => $errorcode, 'friendnum' => $friendnum];
    }

    public function getFollowerOfUser($username)
    {
        $sql = "SELECT * FROM FOLGT
            WHERE USERNAME = '{$username}'";

        $statement = oci_parse($this->conn, $sql);

        oci_execute($statement);

        oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);

        //clean up;
        oci_free_statement($statement);

        return $res;
    }

    public function getHistorieOfUser($username)
    {
        $sql = "SELECT songname, ort, datum
        FROM UserHistorie
        WHERE username = '{$username}'";

        $statement = oci_parse($this->conn, $sql);

        oci_execute($statement);

        oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);

        //clean up;
        oci_free_statement($statement);

        return $res;
    }

}